File: scripts/extract-black.js
import { Jimp } from 'jimp';
import path from 'path';

const COLOR_MAP = {
  white: { r: 255, g: 255, b: 255 },
  gold: { r: 255, g: 215, b: 0 },
  purple: { r: 160, g: 32, b: 240 },
};

async function extractBlack(inputPath, outputPath, colorName = 'white') {
  const targetColor = COLOR_MAP[colorName] || COLOR_MAP.white;
  const image = await Jimp.read(inputPath);
  
  const width = image.bitmap.width;
  const height = image.bitmap.height;
  
  let minX = width, maxX = 0, minY = height, maxY = 0;
  let foundAny = false;

  // Process pixels
  image.scan(0, 0, width, height, function(x, y, idx) {
    const r = this.bitmap.data[idx + 0];
    const g = this.bitmap.data[idx + 1];
    const b = this.bitmap.data[idx + 2];
    const a = this.bitmap.data[idx + 3];

    // Dark enough (R<80, G<80, B<80) and not already transparent
    if (r < 80 && g < 80 && b < 80 && a > 0) {
      this.bitmap.data[idx + 0] = targetColor.r;
      this.bitmap.data[idx + 1] = targetColor.g;
      this.bitmap.data[idx + 2] = targetColor.b;
      this.bitmap.data[idx + 3] = 255;
      
      if (x < minX) minX = x;
      if (x > maxX) maxX = x;
      if (y < minY) minY = y;
      if (y > maxY) maxY = y;
      foundAny = true;
    } else {
      this.bitmap.data[idx + 3] = 0; // Transparent
    }
  });

  if (!foundAny) {
    console.error('No dark pixels found!');
    return false;
  }

  // Crop with 10px padding
  const padding = 10;
  const left = Math.max(0, minX - padding);
  const top = Math.max(0, minY - padding);
  const right = Math.min(width, maxX + padding);
  const bottom = Math.min(height, maxY + padding);
  
  const cropWidth = right - left;
  const cropHeight = bottom - top;

  image.crop(left, top, cropWidth, cropHeight);
  await image.writeAsync(outputPath);
  console.log(`Saved processed image to ${outputPath}`);
  return true;
}

// CLI handling if run directly
if (process.argv[1].endsWith('extract-black.js')) {
  const args = process.argv.slice(2);
  const input = args[0];
  const output = args[1];
  const color = args[2] || 'white';
  
  if (!input || !output) {
    console.log('Usage: node scripts/extract-black.js [input.jpg] [output.png] [color]');
    process.exit(1);
  }

  extractBlack(input, output, color)
    .catch(err => {
      console.error(err);
      process.exit(1);
    });
}

export { extractBlack };

File: scripts/extract-black.py
#!/usr/bin/env python3
# scripts/extract-black.py
# Usage: python scripts/extract-black.py [input.jpg] [output.png] [--color gold|purple|white]
import sys
import os
from PIL import Image, ImageOps

COLOR_MAP = {
    "white": (255, 255, 255),
    "gold": (255, 215, 0),
    "purple": (160, 32, 240),
}

def extract_black(input_path, output_path=None, color_name="white"):
    color_rgb = COLOR_MAP.get(color_name, (255, 255, 255))
    img = Image.open(input_path)
    
    if img.mode != 'RGBA':
        img = img.convert('RGBA')
    
    # Create a new image with transparency
    pixels = img.load()
    width, height = img.size
    
    # Find the bounding box of black/dark gray pixels
    black_pixels = []
    for y in range(height):
        for x in range(width):
            r, g, b, a = pixels[x, y]
            # Check if pixel is dark enough (adjust threshold as needed)
            if r < 80 and g < 80 and b < 80 and a > 0:
                black_pixels.append((x, y))
                pixels[x, y] = (*color_rgb, 255)  # Make the chosen color
            else:
                pixels[x, y] = (0, 0, 0, 0)  # Make transparent
    
    if not black_pixels:
        print("No black pixels found!")
        return
    
    # Find bounding box
    xs = [p[0] for p in black_pixels]
    ys = [p[1] for p in black_pixels]
    min_x, max_x = min(xs), max(xs)
    min_y, max_y = min(ys), max(ys)
    
    # Crop to bounding box with small padding
    padding = 10
    left = max(0, min_x - padding)
    top = max(0, min_y - padding)
    right = min(width, max_x + padding)
    bottom = min(height, max_y + padding)
    
    cropped = img.crop((left, top, right, bottom))
    
    # Save
    if output_path is None:
        base = os.path.splitext(input_path)[0]
        output_path = f"{base}-black.png"
    
    cropped.save(output_path, 'PNG')
    print(f"Saved {output_path}")

if __name__ == '__main__':
    input_path = "scripts/note.jpg"
    output_path = None
    color_name = "white"
    
    for arg in sys.argv[1:]:
        if arg in COLOR_MAP:
            color_name = arg
        elif not output_path:
            input_path = arg
        else:
            output_path = arg
    
    extract_black(input_path, output_path, color_name)